libcxx/include/__memory/compressed_pair.h

```diff
@@ -11,161 +11,79 @@
-      _NOEXCEPT_(__is_nothrow_swappable<_T1>::value&& __is_nothrow_swappable<_T2>::value) {
-    using std::swap;
-    swap(first(), __x.first());
-    swap(second(), __x.second());
-  }
+// The first member is aligned to the alignment of the second member to force padding in front of the compressed pair
+// in case there are members before it.
+//
+// For example:
+// (assuming x86-64 linux)
+// class SomeClass {
+//   uint32_t member1;
+//   _LIBCPP_COMPRESSED_PAIR(uint32_t, member2, uint64_t, member3);
+// }
+//
+// The layout with __compressed_pair is:
+// member1 - offset: 0,  size: 4
+// padding - offset: 4,  size: 4
+// member2 - offset: 8,  size: 4
+// padding - offset: 12, size: 4
+// member3 - offset: 16, size: 8
+//
+// If the [[gnu::aligned]] wasn't there, the layout would instead be:
+// member1 - offset: 0, size: 4
+// member2 - offset: 4, size: 4
+// member3 - offset: 8, size: 8
+
+template <class _ToPad>
+class __compressed_pair_padding {
+  char __padding_[(!is_object<_ToPad>::value || (is_empty<_ToPad>::value && !__libcpp_is_final<_ToPad>::value))
+                      ? 0
+                      : sizeof(_ToPad) - __libcpp_datasizeof<_ToPad>::value];
 };
 
-template <class _T1, class _T2>
-inline _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX14 void
-swap(__compressed_pair<_T1, _T2>& __x, __compressed_pair<_T1, _T2>& __y)
-    _NOEXCEPT_(__is_nothrow_swappable<_T1>::value&& __is_nothrow_swappable<_T2>::value) {
-  __x.swap(__y);
-}
+#  define _LIBCPP_COMPRESSED_PAIR(T1, Name1, T2, Name2)                                                                \
+    [[__gnu__::__aligned__(_LIBCPP_ALIGNOF(T2))]] _LIBCPP_NO_UNIQUE_ADDRESS T1 Name1;                                  \
+    _LIBCPP_NO_UNIQUE_ADDRESS __compressed_pair_padding<T1> Name1##_padding_;                                          \
+    _LIBCPP_NO_UNIQUE_ADDRESS T2 Name2;                                                                                \
+    _LIBCPP_NO_UNIQUE_ADDRESS __compressed_pair_padding<T2> Name2##_padding_
+
+#  define _LIBCPP_COMPRESSED_TRIPLE(T1, Name1, T2, Name2, T3, Name3)                                                   \
+    [[using __gnu__: __aligned__(_LIBCPP_ALIGNOF(T2)),                                                                 \
+      __aligned__(_LIBCPP_ALIGNOF(T3))]] _LIBCPP_NO_UNIQUE_ADDRESS T1 Name1;                                           \
+    _LIBCPP_NO_UNIQUE_ADDRESS __compressed_pair_padding<T1> Name1##padding_;                                           \
+    _LIBCPP_NO_UNIQUE_ADDRESS T2 Name2;                                                                                \
+    _LIBCPP_NO_UNIQUE_ADDRESS __compressed_pair_padding<T2> Name2##padding_;                                           \
+    _LIBCPP_NO_UNIQUE_ADDRESS T3 Name3;                                                                                \
+    _LIBCPP_NO_UNIQUE_ADDRESS __compressed_pair_padding<T3> Name3##padding_
```

Comment:
```suggestion
#  define _LIBCPP_COMPRESSED_TRIPLE(T1, Initializer1, T2, Initializer2, T3, Initializer3)                                                   \
    [[using __gnu__: __aligned__(_LIBCPP_ALIGNOF(T2)),                                                                 \
      __aligned__(_LIBCPP_ALIGNOF(T3))]] _LIBCPP_NO_UNIQUE_ADDRESS T1 Initializer1;                                           \
    _LIBCPP_NO_UNIQUE_ADDRESS __compressed_pair_padding<T1> _LIBCPP_CONCAT(__padding1_, __LINE__);                                           \
    _LIBCPP_NO_UNIQUE_ADDRESS T2 Initializer2;                                                                                \
    _LIBCPP_NO_UNIQUE_ADDRESS __compressed_pair_padding<T2> _LIBCPP_CONCAT(__padding2_, __LINE__);                                           \
    _LIBCPP_NO_UNIQUE_ADDRESS T3 Initializer3;                                                                                \
    _LIBCPP_NO_UNIQUE_ADDRESS __compressed_pair_padding<T3> _LIBCPP_CONCAT(__padding3_, __LINE__)
```

This way, you can do `_LIBCPP_COMPRESSED_TRIPLE(T, foo = 3, ...)` to have member initializers.